keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,992 @@
+
+##### Node Deleted
+
+Release said Node’s allocation from the internal data-structure.
+
+If this Node is the last one using a particular `ClusterCIDRConfig` that has
+been slated for deletion, trigger the deletion flow again (so that the finalizer
+is removed and internal data structures are cleaned up).
+
+##### ClusterCIDRConfig Added
+
+Install a finalizer on the `ClusterCIDRConfig` called "networking.kubernetes.io/cluster-cidr-config-finalizer".
+
+Update internal representation of CIDRs to include the new range. Every failed
+Node Allocation is stored in a queue, that will be tried again with the new
+range by the reconciliation loop.
+
+##### ClusterCIDRConfig Updated
+
+_`CIDR`, and `PerNodeMaskSize` are immutable_
+
+Update any internal state as required.
+
+If the `NodeSelector` changed to include a node that is currently awaiting a
+PodCIDR allocation, it will be allocated by the reconiliation loop. 
+
+If the `NodeSelector` changes to unselect a Node currently using that range,
+there is no change to the Node's allocation. Users will have to manually
+restart/recreated that Node to reset it's `PodCIDR` allocation.
+
+##### ClusterCIDRConfig Deleted
+
+1.  Update internal data structures to mark the range as terminating (so new
+    nodes won't be added to it)
+1.  Search the internal representation of the CIDR range to see if any Nodes are
+    using the range.
+    1.  If there are no nodes using the range, remove the finalizer and cleanup
+        all internal state.
+    1.  If there are nodes using the range, wait for them to be deleted before
+        removing the finalizer and cleaning up. 
+
+### kube-controller-manager
+
+The flag `--cidr-allocator-type` will be amended to include a new type
+"ClusterCIDRConfigAllocator".
+
+The list of current valid types is
+[here](https://github.com/kubernetes/kubernetes/blob/1ff18a9c43f59ffed3b2d266b31e0d696d04eaff/pkg/controller/nodeipam/ipam/cidr_allocator.go#L38).
+
+### Test Plan
+
+#### Unit Tests and Benchmarks
+
+-   Ensure that the controller scales to ~5,000 nodes -- memory usage and
+    reasonable allocation times
+
+#### Integration Tests
+
+-   Verify finalizers and statuses are persisted appropriately
+-   Test watchers
+-   Ensure that the controller handles the feature being disabled and re-enabled:
+    -   Test with some Nodes already having `PodCIDR` allocations
+
+#### End-to-End Tests
+
+-   Run through some sample workflows. Just a few for example:
+    -   Adding a node
+    -   Adding a ClusterCIDRConfig
+    -   Deleting a ClusterCIDRConfig that is in use
+-   Run through the [user stories](#user-stories):
+    -   Expand the ClusterCIDR (existing nodes without alloations are
+        allocated and new nodes also get ranges.
+    -   Use `NodeSelector` to allocate different sized CIDRs to different nodes.
+    -   Create and use discontiguous ranges.
+
+### Graduation Criteria
+
+The initial release of this feature will be as a CRD and custom controller.
```

Comment:
how the custom controller will inherit the flags of the controller manager?

Reply:
Hmm that's a good point -- I was banking on this being a built-in controller running in kube-controller-manager. We're also rely on kube-controller-manager's leader election so this might be a rather tricky situation.

Reply:
I think that this is well encapsulated and the API is relatively simple, so we can avoid the CRD/custom controller , but better ask for permission first to @thockin ;)